namesys/interface.go

```diff
@@ -60,39 +57,152 @@ type NameSystem interface {
 	Publisher
 }
 
-// Result is the return type for Resolver.ResolveAsync.
-type Result struct {
+// ResolveResult is the return type for [Resolver.ResolveAsync].
+type ResolveResult struct {
 	Path path.Path
+	TTL  time.Duration
 	Err  error
 }
 
 // Resolver is an object capable of resolving names.
 type Resolver interface {
-	// Resolve performs a recursive lookup, returning the dereferenced
-	// path.  For example, if ipfs.io has a DNS TXT record pointing to
+	// Resolve performs a recursive lookup, returning the dereferenced path and the TTL.
+	// If the TTL is unknown, then a TTL of 0 is returned. For example, if example.com
+	// has a DNS TXT record pointing to:
+	//
 	//   /ipns/QmatmE9msSfkKxoffpHwNLNKgwZG8eT9Bud6YoPab52vpy
+	//
 	// and there is a DHT IPNS entry for
+	//
 	//   QmatmE9msSfkKxoffpHwNLNKgwZG8eT9Bud6YoPab52vpy
 	//   -> /ipfs/Qmcqtw8FfrVSBaRmbWwHxt3AuySBhJLcvmFYi3Lbc4xnwj
+	//
 	// then
+	//
 	//   Resolve(ctx, "/ipns/ipfs.io")
+	//
 	// will resolve both names, returning
+	//
 	//   /ipfs/Qmcqtw8FfrVSBaRmbWwHxt3AuySBhJLcvmFYi3Lbc4xnwj
 	//
-	// There is a default depth-limit to avoid infinite recursion.  Most
-	// users will be fine with this default limit, but if you need to
-	// adjust the limit you can specify it as an option.
-	Resolve(ctx context.Context, name string, options ...opts.ResolveOpt) (value path.Path, err error)
+	// There is a default depth-limit to avoid infinite recursion. Most users will be fine with
+	// this default limit, but if you need to adjust the limit you can specify it as an option.
+	Resolve(ctx context.Context, name string, options ...ResolveOption) (value path.Path, ttl time.Duration, err error)
+
+	// ResolveAsync performs recursive name lookup, like Resolve, but it returns entries as
+	// they are discovered in the DHT. Each returned result is guaranteed to be "better"
+	// (which usually means newer) than the previous one.
+	ResolveAsync(ctx context.Context, name string, options ...ResolveOption) <-chan ResolveResult
```

Comment:
I'm considering to change the `name` here for the type `path.Path`. The name here is supposed to be a path. This is unrelated to the routing key in the DHT. These are the human readable names. So I think it would make sense.

```suggestion
	// There is a default depth-limit to avoid infinite recursion. Most users will be fine with
	// this default limit, but if you need to adjust the limit you can specify it as an option.
	Resolve(ctx context.Context, name path.Path, options ...ResolveOption) (value path.Path, ttl time.Duration, err error)

	// ResolveAsync performs recursive name lookup, like Resolve, but it returns entries as
	// they are discovered in the DHT. Each returned result is guaranteed to be "better"
	// (which usually means newer) than the previous one.
	ResolveAsync(ctx context.Context, name path.Path, options ...ResolveOption) <-chan ResolveResult
```

Reply:
The issue here is that right now we support things such as `Resolve(ctx, "dnslink.net")` and `Resolve(ctx, "peer-id")`, which means we do not necessarily need a namespace, or a path. Changing that would break some things. It would still be possible to keep Kubo's CLI with the same API with no changes.

What do you think? @lidel

Reply:
Probably better keep `name string` as-is, as allowing paths will complicate implementation and tests.

Currently, the path would effectively always be `/ipns/id`.
But if you allow paths,  we need to decide what happens when someone passes   `/ipns/id/foo/bar`.
Do we keep path remainer and append it to returned path? Do consumers of this library expect that?

Keeping `name string` keeps scope limited to the root identifier, and consumer of the library know they always get path in return for the `id` (and if they had any sub-path, they append themselves).

Reply:
@lidel well, `Resolve` already takes a path and not a `name` per-sea. And it's even more complicated:
- `NewNameSystem` returns a namesys that cleans up the path, gets the name identifier out of it, passes it to the underlying resolver, appends the remaining path.
- `DNSResolver` strips the /ipns prefix, resolves, appends path.
- `IPNSResolver` strips the /ipns prefix, resolves, **does not append path**, assuming it's being used with the `NewNameSystem`. 

So basically they all have slightly different behaviours, but they're "saved" because they're never used by their own. I will keep the name because I agree it's __just__ simpler, but either way we'll always be sanitizing and cleaning it up inside all the functions.

Reply:
My suggestion: make `.Resolve` strictly only accept `[/ipns/]{name}` and have some utility function like what we already have nowadays (`ResolveIPNS`) that does the path handling. But here's where the problem arises: recursion. What if an IPNS name's value is `/ipns/{name}/a/b`? How will we recursively solve that now that we no longer accept paths?

Reply:
Wow, indeed this is worse than I've remembered :see_no_evil:  Thank you for the in-depth context.

It feels like dechnical debt. There was a time when we preferred to design libraries that represent paths as `string`  to reduce dependency hell, but  that has not been the case for a long time (at some point we started  returning them as explicit  Path type and not a `string`).

We already return `path.Path`, so using the same type for input (1) sounds logical.
(2) would make sense if we removed `Path` from all interfaces.